### PR TITLE
replace use of std math in kernels with quda versions

### PR DESCRIPTION
**Pull request by jcosborn**
replace use of std math in kernels with quda versions
allow building without cufft
change some casting

### CODE REVIEW

File: include/targets/hip/shared_memory_helper.h [hunk 1]
```diff
@@ -72,8 +72,11 @@ namespace quda
     /**
        @brief Constructor for SharedMemory object.
     */
-    constexpr SharedMemory() : data(cache(get_offset(target::block_dim()))) { }
+    __device__ __host__ constexpr SharedMemory() : data(cache(get_offset(target::block_dim()))) { }
```
Comment (maddyscientist) [inline include/targets/hip/shared_memory_helper.h]: Why is this needed?

File: include/targets/cuda/shared_memory_helper.h [hunk 2]
```diff
@@ -72,8 +72,11 @@ namespace quda
     /**
        @brief Constructor for SharedMemory object.
     */
-    constexpr SharedMemory() : data(cache(get_offset(target::block_dim()))) { }
+    __device__ __host__ constexpr SharedMemory() : data(cache(get_offset(target::block_dim()))) { }
```
Comment (maddyscientist) [inline include/targets/cuda/shared_memory_helper.h]: Why is this needed?

File: include/targets/generic/mrg32k3a.h [hunk 3]
```diff
@@ -191,14 +192,14 @@ namespace quda
       constexpr double uniform(MRG32k3a &prn)
       {
         double p1 = a12 * (double)(prn.s1[1]) - a13n * (double)(prn.s1[0]);
-        p1 = std::fmod(p1, m1);
+        p1 = fmod(p1, m1);
```
Comment (maddyscientist) [inline include/targets/generic/mrg32k3a.h:195]: Confused here.  There is no `quda::fmod` function in the math_helper, so why making this change?

File: include/targets/cuda/math_helper.cuh [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 #include <cmath>
 #include <target_device.h>
+#include <quda_cuda_api.h>
```
Comment (maddyscientist) [inline include/targets/cuda/math_helper.cuh]: Why did this header need to be added here?

**Review by maddyscientist — COMMENTED** [review]
Looks good James.  I've left a few questions.

Perhaps I forget the context, but what is the goal to allow building without cuFFT on CUDA?  Does the same need to be enabled for the HIP target?

File: include/targets/hip/shared_memory_helper.h (hunk 1 above)
Comment (jcosborn) [reply include/targets/hip/shared_memory_helper.h]: Not sure why I added that.  I've removed it now (21d2cadf0cd2a4c23e4131b6bad0a3e92c9f7dc3).

File: include/targets/cuda/shared_memory_helper.h (hunk 2 above)
Comment (jcosborn) [reply include/targets/cuda/shared_memory_helper.h]: Not sure why I added that.  I've removed it now (21d2cadf0cd2a4c23e4131b6bad0a3e92c9f7dc3).

File: include/targets/generic/mrg32k3a.h (hunk 3 above)
Comment (jcosborn) [reply include/targets/generic/mrg32k3a.h:195]: You're right.  I think SYCL on CUDA was still working without the quda version.  For consistency though we should have the quda version so I've add that in (21d2cadf0cd2a4c23e4131b6bad0a3e92c9f7dc3).

File: include/targets/cuda/math_helper.cuh (hunk 4 above)
Comment (jcosborn) [reply include/targets/cuda/math_helper.cuh]: Not sure why, but there probably was a reason.  I've reverted for now and will revisit if I encounter it again.
21d2cadf0cd2a4c23e4131b6bad0a3e92c9f7dc3

**Comment by jcosborn** [comment]
> Perhaps I forget the context, but what is the goal to allow building without cuFFT on CUDA? Does the same need to be enabled for the HIP target?

This was for testing with LLVM on Polaris.  By default, it didn't find the libraries, and it seemed easier to just disable them than try to point it in the right direction.  Disabling worked fine for blas, but wasn't fully implemented for fft.
I don't think there is a similar need for HIP at the moment.  It could be attempted now for completeness, or just wait until there is a test case.

**Review by maddyscientist — APPROVED** [review]
Approved, but one final question: I see you've restore the `constexpr` but you still have the `__device__ __host__` qualifiers on the shared memory constructor.  Were there needed for some compiler?

**Comment by jcosborn** [comment]
I think you mean in the `sharedMem` accessor?  I just reverted that too.
f9a551aac4397ef6975f59d64d55bc4087398ea1

**Comment by maddyscientist** [comment]
cscs-ci run
